jupyter_server/services/api/handlers.py

```diff
@@ -57,7 +60,46 @@ async def get(self):
         self.finish(json.dumps(model, sort_keys=True))
 
 
+class IdentityHandler(APIHandler):
+    """Get the current user's identity model"""
+
+    @web.authenticated
+    def get(self):
+        permissions_json: str = self.get_argument("permissions", "")
+        bad_permissions_msg = f'permissions should be a JSON dict of {{"resource": ["action",]}}, got {permissions_json!r}'
+        if permissions_json:
+            try:
+                permissions_to_check = json.loads(permissions_json)
+            except ValueError:
+                raise web.HTTPError(400, bad_permissions_msg)
+            if not isinstance(permissions_to_check, dict):
+                raise web.HTTPError(400, bad_permissions_msg)
+        else:
+            permissions_to_check = {}
+
+        permissions: Dict[str, List[str]] = {}
+        user = self.current_user
+
+        for resource, actions in permissions_to_check.items():
+            if (
+                not isinstance(resource, str)
+                or not isinstance(actions, list)
+                or not all(isinstance(action, str) for action in actions)
+            ):
+                raise web.HTTPError(400, bad_permissions_msg)
+
+            allowed = permissions[resource] = []
+            for action in actions:
+                if self.authorizer.is_authorized(self, user=user, resource=resource, action=action):
+                    allowed.append(action)
+
+        identity: IdentityModel = self.identity_provider._get_identity_model(user)
+        identity["permissions"] = permissions
```

Comment:
The return value from the identity provider + the `IdentityModel` type might give the incorrect impression that the identity provider is meant to provide the permissions. I think we should not include the "permissions" on the returned value from the provider (or it's type), and also more clearly document the interplay between the identity provider and the authorizer (i.e. that the username from `get_user` is passed to the authorizer).

Reply:
Makes sense. Do you think checking permissions should be a separate endpoint, or make the identity and permisson models siblings in the current endpoint:

```
identity: IdentityModel
permissions: permissionsModel
```

Reply:
Tough call. On the one hand, having the APIs match the server side architecture/design makes sense, OTOH it might be good from the API callers POV to get all the data in one request / structure. Some thinking out loud:
- You can fetch either without invoking the other (they both rely on auth, but not each other). If either operation is ever "expensive" (slow, or resource use heavy), it might be good to separate the calls from each other. E.g.
  - currently the "permissions" part of this request is opt-in (if no list of permissions are request, no checks are done). This means that if you only want to know about the identity, no permissions checks are done.
  - if the `identity_model` needs to consult an external resource, it might be good for it not to be invoked for every permissions check. Depending on how the identity model is designed client side, you might end up with more frequent permissions check, while the identity model is only fetched once per user. Especially since you might want to use the API to only check only a filtered subset of permissions (but each check will include the full identity model).

Maybe adding a separate handler for permissions makes sense? Even if we do, keeping the "opt-in" of permissions check in the same request as the identity might be convenient? I'm not 100% sure either direction.

Reply:
I see I forgot how the code worked over the weekend: 
- The identity model is needed for permissions checks, so it comes "for free" with every check.
- As such, it might be good to encourage as few separate calls for permissions as possible (in case either operation is expensive).
In light of that, keeping them as one API request probably makes sense, given that the permissions checks are opt-in. Making the models siblings might make sense, but if any given permission model always have only one unique owning identity, it probably makes sense to keep it as is.

Reply:
Went with `/api/me` returning

```
identity: IdentityModel
permissions: PermissionsModel
```

for now. I do think it's preferable for this to be a single endpoint, but I really don't have a strong preference for whether 'permissions' are a sibling to 'username' or 'identity'.